src/effects/VST3/VST3Utils.h

```diff
@@ -52,7 +56,73 @@ class VST3Utils final
    //a human-readable form.
    static wxString MakeAutomationParameterKey(const Steinberg::Vst::ParameterInfo& info);
 
-   //Attempts to extract a unique VST3 parameter id from the key.
-   //Returns true on success.
-   static bool ParseAutomationParameterKey(const wxString& key, Steinberg::Vst::ParamID& paramId);
 };
+
+struct EffectSettings;
+
+struct VST3EffectSettings
+{
+   std::unordered_map<Steinberg::Vst::ParamID, Steinberg::Vst::ParamValue> mValues;
+};
+
+struct VST3Wrapper
+{
+   // For the time being, here we have only members that are needed
+   // to iterate parameters and extract preset state
+   //
+   Steinberg::IPtr<Steinberg::Vst::IEditController> mEditController;
+   Steinberg::IPtr<Steinberg::Vst::IComponent>      mEffectComponent;
+
+   using ParameterInfo = Steinberg::Vst::ParameterInfo;
+
+   using ParameterVisitor = std::function< bool(const ParameterInfo& pi) >;
+
+   // made mutable because it is cached once on demand
+   mutable std::vector<ParameterInfo> mParameterInfos;
+
+   // Visit parameters and keep doing so as long as visitor returns true.
+   //
+   // Return false if the parameters are not yet available
+   //
+   bool ForEachParameter(ParameterVisitor visitor) const;
+
+   // Stop visiting as soon as the visitor returns true, and return true in that case;
+   // otherwise if all visits returned false,  return false.
+   //
+   bool AnyOf(ParameterVisitor visitor) const;
+
+   const std::vector<ParameterInfo>* getParameterInfos() const;
```

Comment:
Pointer or reference as return type?  I would prefer reference.  It's a conventional way of saying, null isn't possible, and making it unnecessary to specify that with a `//! @post` comment.

But otherwise, seeing just this declaration, I as a possible caller of this function, would be left in doubt whether null checks on all the uses of this function should be done or not.

Reply:
Now I see the lines in the function:

```
     if (mEditController == nullptr)
         return nullptr;
```

suggesting that sometimes it's necessary to return null because the effect plugin can't be completely initialized yet.  Is that right?  Then, a comment in the header should explain.

Reply:
That's right, it is for the case when mEditController could not be initialized.
Added comment for it.